TM-41 user list and user details page

TomsBemis opened the pull request:
1) User list page navigation link
<img width="637" alt="image" src="https://github.com/user-attachments/assets/0f9b353d-0d16-43bf-b459-199fe5e849cf">

2) User list page
<img width="638" alt="image" src="https://github.com/user-attachments/assets/361f3081-5ba3-4727-989e-b60ad5a3a67c">

3) User details page with role editing form for admin
<img width="641" alt="image" src="https://github.com/user-attachments/assets/e4fda811-3fe8-40b0-82af-08c9e0fef0b9">

4) User details page without role editing form for user
<img width="640" alt="image" src="https://github.com/user-attachments/assets/de4d771d-3f96-4ee2-a876-3e3d07404319">

TomsBemis committed:
TM-41 User list page and details page

 - Refactored interfaces organizing them in user, option, auth, task models;
 - New user service with methods for getting all users or one user by id;
 - New user attribute - role.

TomsBemis committed:
TM-41 Token interceptor changed append access token method

TomsBemis committed:
TM-41 Task list component always fetches tasks from service on initialization

TomsBemis committed:
TM-41 Logout not deleting cookies fix

TomsBemis committed:
Merge branch 'develop' into TM-41-user-list-and-user-details-page

TomsBemis committed:
TM-41 Merge commit fix

TomsBemis committed:
TM-41 Fixed return type of auth service logout method

TomsBemis committed:
TM-41 Edit user roles method

TomsBemis committed:
TM-41 Navigation link to users list page for admin and login logic moved from login component to auth service

TomsBemis committed:
TM-41 User details page link and restrictions

 - Moved user details page restrictions to user details component (user details and user list routes clash)
 - Removed unnecessary console logs
 - Navigation link to user details page on navbar
 - Auth guard checks with auth service
 - Guards refactored to separate directory

wanpii commented on src/app/guards/userRole.guard.ts:
```diff
@@ -0,0 +1,38 @@
+
+    // Check if user has appropriate role
+    if(whitelist) {
+      for (let i = 0; i < criteriaRoles.length; i++) {
```
you can use js array functions like forEach and includes to minimise the amount of code

wanpii commented on src/app/guards/userRole.guard.ts:
```diff
@@ -0,0 +1,38 @@
+    }
+    else {
+      for (let i = 0; i < criteriaRoles.length; i++) {
+        if(authenticatedUser.role.value == criteriaRoles[i].value) {
```
same here

wanpii commented on src/app/shared/nav/nav.component.html:
```diff
@@ -11,6 +11,11 @@
             <li class="nav-item">
               <a class="nav-link" routerLink="/tasks/create">Create new task</a>
             </li>
+            @if(currentUser.value?.role?.value == "ADMIN") {
```
little bit awkward having currentUser.value.role.value value twice 😆

TomsBemis replied:
Will change when I figure out how to track a local variable in template

wanpii commented on src/app/shared/nav/nav.component.html:
```diff
@@ -23,7 +28,7 @@
             <li class="nav-item">
               <div class="nav-link">
-                Welcome {{currentUser.getValue()?.firstName + ' ' + currentUser.getValue()?.lastName}}
+                <a class="nav-link" [routerLink]="['/users', currentUser.getValue()?.id]">Welcome {{currentUser.getValue()?.firstName + ' ' + currentUser.getValue()?.lastName}}</a>
```
you're async'ing the currentUser$ observable above, you can declare it as a variable inside the template by adding as <variable_name> something like this:

if (currentUser$ | async as currentUser) then you can use this currentUser variable in the template and directly access it's fields, like currentUser.id etc.

TomsBemis replied:
I could only get this to work in for loops
![image](https://github.com/user-attachments/assets/3dc4e54b-8346-4e32-93c1-523c75dcb3df)

wanpii commented on src/app/users/user-list/user-list.component.ts:
```diff
@@ -0,0 +1,38 @@
+
+    // Check if logged in user is admin, or if not then compare ids
+    let loggedInUser = this.authService.currentUserSubject.getValue();
+    if(!loggedInUser || loggedInUser.role.value != "ADMIN") throw Error("Only users with administrator priviledges have access.");
```
you could add this logic in the page guard as it's related to permissions of accessing the page

TomsBemis replied:
I decided on my implementation because the route /users/:userId should be accessible for the user itself so that they can view their own user details page. But putting a guard that checks if the user is admin on the route /users, makes the route /users/:userId not accessible even if the user is checking their own details. Both routes start with /users and admin guard gets called by default.

wanpii replied:
you could check in the guard if the user is the same one we are accessing or if the role is admin then we let the user access the page, what do you think?

wanpii commented on src/app/users/user.service.ts:
```diff
@@ -0,0 +1,53 @@
+export class UserService{
+    
+    public usersSubject = new BehaviorSubject<User[]>([]);
+    public usersObservable$ = this.usersSubject.asObservable();
```
you can remove the Observable name, just users$ the $ sign signifies that it's an observable

wanpii commented on src/app/users/user.service.ts:
```diff
@@ -0,0 +1,53 @@
+                    (userA, userB) => {
+                        let userFullNameA = userA.firstName+' '+userA.lastName;
+                        let userFullNameB = userB.firstName+' '+userB.lastName;
+                        return (userFullNameA < userFullNameB) ? -1 : (userFullNameA > userFullNameB) ? 1 : 0
```
there should be a less confusing way to write this maybe?

TomsBemis committed:
TM-41 Pull request 13 For loop replaced with array functions

Reviewer comment: https://github.com/TomsBemis/task-manager-frontend/pull/18#discussion_r1760994672

TomsBemis committed:
TM-41 Pull request 18 Requested changes 1

Renamed observable variable names and made user sorting logic more readable

TomsBemis committed:
TM-41 Pull Request 18 Tracking a local variable in template
